File: cs/Pragmastat/Sample.cs
using System.Globalization;
using System.Text;
using JetBrains.Annotations;
using Pragmastat.Exceptions;
using Pragmastat.Internal;
using Pragmastat.Metrology;

namespace Pragmastat;

public class Sample
{
  public IReadOnlyList<double> Values { get; }
  public IReadOnlyList<double>? Weights { get; }
  public double TotalWeight { get; }
  public bool IsWeighted { get; }
  public MeasurementUnit Unit { get; }

  private readonly Lazy<(IReadOnlyList<double> SortedValues, IReadOnlyList<double>? SortedWeights)> lazySortedData;

  public IReadOnlyList<double> SortedValues => lazySortedData.Value.SortedValues;
  public IReadOnlyList<double>? SortedWeights => lazySortedData.Value.SortedWeights;

  /// <summary>
  /// Sample size (always positive)
  /// </summary>
  public int Size => Values.Count;

  /// <summary>
  /// Kish's Effective Sample Size
  /// </summary>
  public double WeightedSize { get; }

  public Sample(params double[] values) : this((IReadOnlyList<double>)values, null, null)
  {
  }

  public Sample(params int[] values) : this(Array.ConvertAll(values, x => (double)x), null, null)
  {
  }

  internal Subject? ValidationSubject { get; }

  public Sample(IReadOnlyList<double> values, MeasurementUnit? unit = null, Subject? validationSubject = null)
  {
    // Validate validity assumption at construction time
    if (values == null)
      throw new ArgumentNullException(nameof(values), "values can't be null");
    ValidationSubject = validationSubject ?? Subject.X;
    if (values.Count == 0)
      throw AssumptionException.Validity(ValidationSubject.Value);
    foreach (var value in values)
    {
      if (!value.IsFinite())
        throw AssumptionException.Validity(ValidationSubject.Value);
    }

    Values = values;
    Unit = unit ?? MeasurementUnit.Number;
    Weights = null;
    TotalWeight = 1.0;
    WeightedSize = values.Count;
    IsWeighted = false;

    lazySortedData = new Lazy<(IReadOnlyList<double>, IReadOnlyList<double>?)>(() =>
      (IsSorted(Values) ? Values : Values.CopyToArrayAndSort(), null));
  }

  public Sample(IReadOnlyList<double> values, IReadOnlyList<double> weights, MeasurementUnit? measurementUnit = null, Subject? validationSubject = null)
  {
    // Validate validity assumption at construction time
    if (values == null)
      throw new ArgumentNullException(nameof(values), "values can't be null");
    ValidationSubject = validationSubject ?? Subject.X;
    if (values.Count == 0)
      throw AssumptionException.Validity(ValidationSubject.Value);
    foreach (var value in values)
    {
      if (!value.IsFinite())
        throw AssumptionException.Validity(ValidationSubject.Value);
    }

    Assertion.NotNullOrEmpty(nameof(weights), weights);
    if (values.Count != weights.Count)
      throw new ArgumentException(
        $"{nameof(weights)} should have the same number of elements as {nameof(values)}",
        nameof(weights));

    double totalWeight = 0, maxWeight = double.MinValue, minWeight = double.MaxValue;
    double totalWeightSquared = 0;
    foreach (double weight in weights)
    {
      totalWeight += weight;
      totalWeightSquared += weight.Sqr();
      maxWeight = Math.Max(maxWeight, weight);
      minWeight = Math.Min(minWeight, weight);
    }

    if (minWeight < 0)
      throw new ArgumentOutOfRangeException(nameof(weights),
        $"All weights in {nameof(weights)} should be non-negative");
    if (totalWeight < 1e-9)
      throw new ArgumentException(nameof(weights),
        $"The sum of all elements from {nameof(weights)} should be positive");

    Values = values;
    Weights = weights;
    Unit = measurementUnit ?? MeasurementUnit.Number;
    TotalWeight = totalWeight;
    WeightedSize = totalWeight.Sqr() / totalWeightSquared;
    IsWeighted = true;

    lazySortedData = new Lazy<(IReadOnlyList<double>, IReadOnlyList<double>?)>(() =>
    {
      if (IsSorted(Values))
        return (Values, Weights);

      double[] sortedValues = Values.CopyToArray();
      double[] sortedWeights = Weights!.CopyToArray();
      Array.Sort(sortedValues, sortedWeights);

      return (sortedValues, sortedWeights);
    });
  }

  [PublicAPI]
  public Sample(IEnumerable<int> values, MeasurementUnit? unit = null)
    : this(values.Select(x => (double)x).ToList(), unit)
  {
  }

  public Sample(IEnumerable<long> values, MeasurementUnit? unit = null)
    : this(values.Select(x => (double)x).ToList(), unit)
  {
  }

  public Sample ConvertTo(MeasurementUnit target)
  {
    if (!Unit.IsCompatible(target))
      throw new UnitMismatchException(Unit, target);
    if (Unit == target) return this;
    double factor = MeasurementUnit.ConversionFactor(Unit, target);
    double[] converted = new double[Size];
    for (int i = 0; i < Size; i++)
      converted[i] = Values[i] * factor;
    return IsWeighted
      ? new Sample(converted, Weights!, target)
      : new Sample(converted, target);
  }

  public Sample Concat(Sample sample)
  {
    if (!Unit.IsCompatible(sample.Unit))
      throw new UnitMismatchException(Unit, sample.Unit);

    var target = MeasurementUnit.Finer(Unit, sample.Unit);

    IEnumerable<double> GetValues(Sample s)
    {
      if (s.Unit == target)
        return s.Values;
      double factor = MeasurementUnit.ConversionFactor(s.Unit, target);
      return s.Values.Select(x => x * factor);
    }

    double[] UniformWeights(int count)
    {
      double w = 1.0 / count;
      double[] result = new double[count];
      for (int i = 0; i < count; i++)
        result[i] = w;
      return result;
    }

    var values1 = GetValues(this);
    var values2 = GetValues(sample);
    var weights1 = Weights ?? UniformWeights(Size);
    var weights2 = sample.Weights ?? UniformWeights(sample.Size);
    return new Sample(values1.Concat(values2).ToList(), weights1.Concat(weights2).ToList(), target);
  }

  public override string ToString()
  {
    var builder = new StringBuilder();
    builder.Append('[');
    for (int i = 0; i < Values.Count; i++)
    {
      if (i != 0) builder.Append(',');
      builder.Append(Values[i].ToString("G", CultureInfo.InvariantCulture));
    }
    builder.Append(']');
    if (Unit.Abbreviation.Length > 0)
      builder.Append(Unit.Abbreviation);
    return builder.ToString();
  }

  public static Sample operator *(Sample sample, double value)
  {
    double[] values = new double[sample.Size];
    for (int i = 0; i < sample.Size; i++)
      values[i] = sample.Values[i] * value;
    return sample.IsWeighted ? new Sample(values, sample.Weights!, sample.Unit) : new Sample(values, sample.Unit);
  }

  public static Sample operator /(Sample sample, double value)
  {
    double[] values = new double[sample.Size];
    for (int i = 0; i < sample.Size; i++)
      values[i] = sample.Values[i] / value;
    return sample.IsWeighted ? new Sample(values, sample.Weights!, sample.Unit) : new Sample(values, sample.Unit);
  }

  public static Sample operator +(Sample sample, double value)
  {
    double[] values = new double[sample.Size];
    for (int i = 0; i < sample.Size; i++)
      values[i] = sample.Values[i] + value;
    return sample.IsWeighted ? new Sample(values, sample.Weights!, sample.Unit) : new Sample(values, sample.Unit);
  }

  public static Sample operator -(Sample sample, double value)
  {
    double[] values = new double[sample.Size];
    for (int i = 0; i < sample.Size; i++)
      values[i] = sample.Values[i] - value;
    return sample.IsWeighted ? new Sample(values, sample.Weights!, sample.Unit) : new Sample(values, sample.Unit);
  }

  public static Sample operator *(double value, Sample sample) => sample * value;
  public static Sample operator +(double value, Sample sample) => sample + value;

  public static Sample operator *(Sample sample, int value) => sample * (double)value;
  public static Sample operator /(Sample sample, int value) => sample / (double)value;
  public static Sample operator +(Sample sample, int value) => sample + (double)value;
  public static Sample operator -(Sample sample, int value) => sample - (double)value;

  public static Sample operator *(int value, Sample sample) => sample * value;
  public static Sample operator +(int value, Sample sample) => sample + value;

  private static bool IsSorted(IReadOnlyList<double> list)
  {
    for (int i = 0; i < list.Count - 1; i++)
      if (list[i] > list[i + 1])
        return false;
    return true;
  }

  public Measurement Min() => SortedValues[0].WithUnit(Unit);
  public Measurement Max() => SortedValues[SortedValues.Count - 1].WithUnit(Unit);

  public Sample Log()
  {
    double[] logValues = new double[Size];
    for (int i = 0; i < Size; i++)
    {
      if (Values[i] <= 0)
        throw AssumptionException.Positivity(ValidationSubject ?? Subject.X);
      logValues[i] = Math.Log(Values[i]);
    }
    return IsWeighted
      ? new Sample(logValues, Weights!, MeasurementUnit.Number)
      : new Sample(logValues, MeasurementUnit.Number);
  }
}

File: py/pragmastat/sample.py
from __future__ import annotations

from functools import cached_property
from typing import TYPE_CHECKING, Sequence

import numpy as np

from .assumptions import AssumptionError, check_validity
from .measurement_unit import NUMBER_UNIT, MeasurementUnit

if TYPE_CHECKING:
    from numpy.typing import NDArray


class Sample:
    """Wraps values with optional weights and a measurement unit."""

    def __init__(
        self,
        values: Sequence[float] | NDArray,
        weights: Sequence[float] | NDArray | None = None,
        unit: MeasurementUnit | None = None,
    ) -> None:
        if unit is None:
            unit = NUMBER_UNIT

        arr = np.array(values, dtype=np.float64)
        # Construction can't know which argument position this sample fills, so
        # construction-time validity errors are always reported under subject "x".
        check_validity(arr, "x")

        _freeze_array(arr)
        self._values: NDArray = arr
        self._unit: MeasurementUnit = unit

        if weights is not None:
            w = np.array(weights, dtype=np.float64)
            if len(w) != len(arr):
                msg = f"weights length ({len(w)}) must match values length ({len(arr)})"
                raise ValueError(msg)
            if np.any(w < 0):
                msg = "all weights must be non-negative"
                raise ValueError(msg)
            total = float(np.sum(w))
            if total < 1e-9:
                msg = "total weight must be positive"
                raise ValueError(msg)
            _freeze_array(w)
            self._weights: NDArray | None = w
            self._is_weighted = True
            self._total_weight = total
            self._weighted_size = float((total * total) / np.sum(w * w))
        else:
            self._weights = None
            self._is_weighted = False
            self._total_weight = 1.0
            self._weighted_size = float(len(arr))

    @property
    def values(self) -> NDArray:
        return self._values

    @property
    def weights(self) -> NDArray | None:
        return self._weights

    @property
    def unit(self) -> MeasurementUnit:
        return self._unit

    @property
    def size(self) -> int:
        return len(self._values)

    @property
    def is_weighted(self) -> bool:
        return self._is_weighted

    @property
    def total_weight(self) -> float:
        return self._total_weight

    @property
    def weighted_size(self) -> float:
        return self._weighted_size

    @cached_property
    def sorted_values(self) -> NDArray:
        return _freeze_array(np.sort(self._values))

    def convert_to(self, target: MeasurementUnit) -> Sample:
        """Converts the sample to a different (compatible) unit."""
        if not self._unit.is_compatible(target):
            msg = f"can't convert {self._unit.full_name} to {target.full_name}"
            raise AssumptionError(msg)
        if self._unit == target:
            return self
        factor = MeasurementUnit.conversion_factor(self._unit, target)
        converted = self._values * factor
        return _build_sample(
            converted,
            target,
            self._is_weighted,
            self._total_weight,
            self._weighted_size,
            self._weights.copy() if self._weights is not None else None,
        )

    # --- Comparison and display ---

    # Unhashable by design: __eq__ compares float array contents element-wise,
    # and no hash can be both cheap and consistent with that value equality.
    __hash__ = None

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Sample):
            return NotImplemented
        return (
            np.array_equal(self._values, other._values)
            and self._unit == other._unit
            and self._is_weighted == other._is_weighted
            and (
                (self._weights is None and other._weights is None)
                or (
                    self._weights is not None
                    and other._weights is not None
                    and np.array_equal(self._weights, other._weights)
                )
            )
        )

    def __repr__(self) -> str:
        if self._is_weighted:
            return f"Sample(size={self.size}, unit={self._unit.id!r}, weighted=True)"
        return f"Sample(size={self.size}, unit={self._unit.id!r})"

    # --- Arithmetic operators ---

    def __mul__(self, scalar: float) -> Sample:
        if not isinstance(scalar, (int, float)):
            return NotImplemented
        return _build_sample(
            self._values * scalar,
            self._unit,
            self._is_weighted,
            self._total_weight,
            self._weighted_size,
            self._weights,
        )

    def __rmul__(self, scalar: float) -> Sample:
        return self.__mul__(scalar)

    def __add__(self, scalar: float) -> Sample:
        if not isinstance(scalar, (int, float)):
            return NotImplemented
        return _build_sample(
            self._values + scalar,
            self._unit,
            self._is_weighted,
            self._total_weight,
            self._weighted_size,
            self._weights,
        )

    def __radd__(self, scalar: float) -> Sample:
        return self.__add__(scalar)

    def __sub__(self, scalar: float) -> Sample:
        if not isinstance(scalar, (int, float)):
            return NotImplemented
        return self.__add__(-scalar)


def _build_sample(  # noqa: PLR0913
    values: NDArray,
    unit: MeasurementUnit,
    is_weighted: bool,
    total_weight: float,
    weighted_size: float,
    weights: NDArray | None,
) -> Sample:
    """Module-level factory that bypasses validation (values already checked)."""
    obj = Sample.__new__(Sample)
    obj._values = _freeze_array(values)  # noqa: SLF001
    obj._unit = unit  # noqa: SLF001
    obj._is_weighted = is_weighted  # noqa: SLF001
    obj._total_weight = total_weight  # noqa: SLF001
    obj._weighted_size = weighted_size  # noqa: SLF001
    obj._weights = _freeze_array(weights) if weights is not None else None  # noqa: SLF001
    return obj


def _freeze_array(array: NDArray) -> NDArray:
    """Marks an array as read-only so cached sorted values cannot go stale."""
    array.setflags(write=False)
    return array


def _check_non_weighted(name: str, s: Sample) -> None:
    """Raises AssumptionError if the sample is weighted."""
    if s.is_weighted:
        msg = f"weighted samples are not supported for {name}"
        raise AssumptionError(msg)


def _check_compatible_units(a: Sample, b: Sample) -> None:
    """Raises AssumptionError if two samples have incompatible units."""
    if not a.unit.is_compatible(b.unit):
        msg = f"can't convert {a.unit.full_name} to {b.unit.full_name}"
        raise AssumptionError(msg)


def _convert_to_finer(a: Sample, b: Sample) -> tuple[Sample, Sample]:
    """Converts both samples to the finer unit."""
    if a.unit == b.unit:
        return a, b
    target = MeasurementUnit.finer(a.unit, b.unit)
    return a.convert_to(target), b.convert_to(target)


def _prepare_pair(x: Sample, y: Sample) -> tuple[Sample, Sample]:
    """Prepare two samples: check unit compatibility, convert to finer unit.

    The error "subject" (x vs y) is determined positionally by the validating
    function, not stored on the Sample, so no relabeling happens here.

    When both samples already share a unit, they are returned unchanged, so any
    warm sorted-value caches survive. When a unit conversion is required,
    ``convert_to`` builds a fresh Sample with scaled values; that new Sample
    starts with a cold cache (the sorted view is NOT carried over).
    """
    _check_compatible_units(x, y)
    return _convert_to_finer(x, y)
